pdns/recursordist/rec-zonetocache.cc

```diff
@@ -190,20 +210,130 @@ static std::vector<std::string> getURL(const RecZoneToCache::Config& config)
+{
+  DNSResourceRecord drr;
+  ZoneParserTNG zpt(lines, d_zone);
+  zpt.setMaxGenerateSteps(1);
+  zpt.setMaxIncludes(0);
+
+  while (zpt.get(drr)) {
+    DNSRecord dr(drr);
+    if (config.d_zonemd != pdns::ZoneMD::Config::Ignore) {
+      zonemd.readRecord(dr);
+    }
+    parseDRForCache(dr);
+  }
+  if (config.d_zonemd != pdns::ZoneMD::Config::Ignore) {
+    bool validationDone, validationSuccess;
+    zonemd.verify(validationDone, validationSuccess);
+    d_log->info("ZONEMD digest validation", "validationDone", Logging::Loggable(validationDone),
+                "validationSuccess", Logging::Loggable(validationSuccess));
+    if (!validationDone) {
+      return pdns::ZoneMD::Result::NoValidationDone;
+    }
+    if (!validationSuccess) {
+      return pdns::ZoneMD::Result::ValidationFailure;
+    }
+  }
+  return pdns::ZoneMD::Result::OK;
+}
+
+vState ZoneData::dnssecValidate(pdns::ZoneMD& zonemd, size_t& zonemdCount) const
+{
+  zonemdCount = 0;
+
+  SyncRes sr({d_now, 0});
+  sr.setDoDNSSEC(true);
+  sr.setDNSSECValidationRequested(true);
+
+  dsmap_t dsmap; // Actually a set
+  vState dsState = sr.getDSRecords(d_zone, dsmap, false, 0);
+  if (dsState != vState::Secure) {
+    return dsState;
+  }
+
+  skeyset_t dnsKeys;
+  sortedRecords_t records;
+  if (zonemd.getDNSKEYs().size() == 0) {
+    return vState::BogusUnableToGetDNSKEYs;
+  }
+  for (const auto& key : zonemd.getDNSKEYs()) {
+    dnsKeys.emplace(key);
+    records.emplace(key);
+  }
+
+  skeyset_t validKeys;
+  vState dnsKeyState = validateDNSKeysAgainstDS(d_now, d_zone, dsmap, dnsKeys, records, zonemd.getRRSIGs(), validKeys);
+  if (dnsKeyState != vState::Secure) {
+    return dnsKeyState;
+  }
+
+  if (validKeys.size() == 0) {
+    return vState::BogusNoValidDNSKEY;
+  }
+
+  auto zonemdRecords = zonemd.getZONEMDs();
+  zonemdCount = zonemdRecords.size();
+
+  // De we need to do a denial validation?
+  if (zonemdCount == 0) {
+    const auto& nsecs = zonemd.getNSECs();
+    const auto& nsec3s = zonemd.getNSEC3s();
+    cspmap_t csp;
+
+    vState nsecValidationStatus;
+    if (nsecs.records.size() > 0 && nsecs.signatures.size() > 0) {
+      nsecValidationStatus = validateWithKeySet(d_now, d_zone, nsecs.records, nsecs.signatures, validKeys);
+      csp.emplace(std::make_pair(d_zone, QType::NSEC), nsecs);
+    }
+    else if (nsec3s.records.size() > 0 && nsec3s.signatures.size() > 0) {
+      nsecValidationStatus = validateWithKeySet(d_now, d_zone, nsec3s.records, nsec3s.signatures, validKeys);
+      csp.emplace(std::make_pair(d_zone, QType::NSEC3), nsec3s);
+    }
+    else {
+      d_log->info("No NSEC(3) records and/or RRSIGS found to deny ZONEMD");
+      return vState::BogusInvalidDenial;
+    }
+
+    if (nsecValidationStatus != vState::Secure) {
+      d_log->info("zone NSEC(3) record does no validate");
+      return nsecValidationStatus;
+    }
+    auto denial = getDenial(csp, d_zone, QType::ZONEMD, false, false, true);
```

Comment:
So we are passing all the NSEC(3)s present in the zone to `getDenial`, right? Since the zone md record has to be at the apex of the zone, and we are pretty sure that there is at least a SOA record at the apex, perhaps we could be a bit smarter than that?

Reply:
NO, I already filter on aped in the code collecting the records. That is not correct for NSEC3, working on that.

Reply:
NSEC3 case is now handled as well